File: flutter_readium/android/src/main/kotlin/dk/nota/flutter_readium/models/GuidedNavigationDocument.kt
package dk.nota.flutter_readium.models

import org.json.JSONArray
import org.json.JSONObject
import org.readium.r2.shared.publication.Link
import org.readium.r2.shared.util.Url
import org.readium.r2.shared.util.mediatype.MediaType

/**
 * A Readium Guided Navigation Document, describing a structured sequence of
 * media-aligned navigation steps for a publication.
 *
 * See https://readium.org/guided-navigation/schema/document.schema.json
 */
data class GuidedNavigationDocument(
    /** The ordered list of guided navigation objects. Contains at least one entry. */
    val guided: List<GuidedNavigationObject>,
    /**
     * Optional cross-references to related resources, using the Readium Web
     * Publication Manifest link schema.
     */
    val links: List<Link> = emptyList(),
) {
    fun toJSON(): JSONObject =
        JSONObject().apply {
            if (links.isNotEmpty()) {
                put("links", JSONArray(links.map { it.toJSON() }))
            }
            put("guided", JSONArray(guided.map { it.toJSON() }))
        }

    fun toJSONString(): String = toJSON().toString()

    companion object {
        fun fromJSON(json: JSONObject?): GuidedNavigationDocument? {
            if (json == null) return null

            val linksArray = json.optJSONArray("links")
            val links =
                buildList {
                    if (linksArray != null) {
                        for (i in 0 until linksArray.length()) {
                            linksArray
                                .optJSONObject(i)
                                ?.let { Link.fromJSON(it) }
                                ?.let { add(it) }
                        }
                    }
                }

            val guided = GuidedNavigationObject.fromJSONArray(json.optJSONArray("guided"))
            if (guided.isEmpty()) return null

            return GuidedNavigationDocument(guided = guided, links = links)
        }

        fun fromJSON(jsonString: String): GuidedNavigationDocument? = fromJSON(JSONObject(jsonString))
    }
}

val guidedNavigationMediaType = MediaType("application/guided-navigation+json")

/**
 * Converts this document to a list of [FlutterMediaOverlay] by flattening all
 * [GuidedNavigationObject]s that carry both an [audioref] and a [textref], then
 * grouping the resulting items by their (audioFile, textFile) pair.
 *
 * @param position    Reading-order position (1-based) shared by all generated items.
 * @param tocHref     ToC href to attach to every item, or null if unknown.
 * @param title       Fallback chapter/section title. Overridden by each object's own text if present.
 * @param readiumOrderItemDuration  Total duration of the reading-order item, used for progression calculations.
 */
fun GuidedNavigationDocument.toMediaOverlays(
    position: Int = 0,
    tocHref: Url? = null,
    title: String = "",
    readiumOrderItemDuration: Double = 0.0,
): List<FlutterMediaOverlay> {
    fun GuidedNavigationObject.flatten(): List<FlutterMediaOverlayItem> {
        val items = mutableListOf<FlutterMediaOverlayItem>()
        if (audioref != null && textref != null) {
            items +=
                FlutterMediaOverlayItem(
                    audio = audioref,
                    text = textref,
                    position = position,
                    tocHref = tocHref,
                    title = title,
                    readingOrderItemDuration = readiumOrderItemDuration,
                )
        }
        children.forEach { items += it.flatten() }
        return items
    }

    val allItems = guided.flatMap { it.flatten() }

    // Group by (audioFile, textFile), preserving insertion order.
    val orderedKeys = mutableListOf<Pair<String, String>>()
    val itemsByKey = linkedMapOf<Pair<String, String>, MutableList<FlutterMediaOverlayItem>>()
    for (item in allItems) {
        val key = item.audioFile to item.textFile
        itemsByKey
            .getOrPut(key) {
                orderedKeys.add(key)
                mutableListOf()
            }.add(item)
    }

    return orderedKeys.map { key -> FlutterMediaOverlay(itemsByKey.getValue(key)) }
}

File: flutter_readium/ios/flutter_readium/Sources/flutter_readium/model/guided-nav/GuidedNavigationDocument.swift
import Foundation
import ReadiumShared

/// A Readium Guided Navigation Document, describing a structured sequence of
/// media-aligned navigation steps for a publication.
///
/// See https://readium.org/guided-navigation/schema/document.schema.json
struct GuidedNavigationDocument: Equatable {
  /// Optional cross-references to related resources, using the Readium Web
  /// Publication Manifest link schema.
  let links: [Link]

  /// The ordered list of guided navigation objects. Contains at least one entry.
  let guided: [GuidedNavigationObject]

  func toJson() -> [String: Any] {
    var res: [String: Any] = [:]
    if !links.isEmpty { res["links"] = links.map { $0.jsonValue } }
    res["guided"] = guided.map { $0.toJson() }
    return res
  }

  func toJsonString(pretty: Bool = false) -> String? {
    let options: JSONSerialization.WritingOptions = pretty ? [.prettyPrinted] : []
    guard let data = try? JSONSerialization.data(withJSONObject: toJson(), options: options) else {
      return nil
    }
    return String(data: data, encoding: .utf8)
  }

  static func fromJson(_ json: [String: Any]?) -> GuidedNavigationDocument? {
    guard let json = json else { return nil }

    let linksJson = json["links"] as? [[String: Any]] ?? []
    let links = linksJson.compactMap { try? Link(json: JSONValue($0)) }

    let guidedJson = json["guided"] as? [[String: Any]] ?? []
    let guided = guidedJson.compactMap { GuidedNavigationObject.fromJson($0) }

    guard !guided.isEmpty else { return nil }

    return GuidedNavigationDocument(links: links, guided: guided)
  }

  static func fromJson(_ jsonString: String) throws -> GuidedNavigationDocument? {
    let data = Data(jsonString.utf8)
    let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
    return fromJson(object)
  }

  /// Converts this document to a list of `FlutterMediaOverlay` by flattening all
  /// `GuidedNavigationObject`s that carry both an `audioref` and a `textref`, then
  /// grouping the resulting items by their (audioFile, textFile) pair.
  ///
  /// - Parameters:
  ///   - position: Reading-order position (1-based) shared by all generated items.
  ///   - tocHref: ToC href to attach to every item, or nil if unknown.
  ///   - title: Chapter/section title, supplied by the caller from the publication ToC.
  ///   - readingOrderDuration: Total duration of the reading-order item, used for progression calculations.
  func toMediaOverlays(
    atPosition position: Int = 0,
    atTocHref tocHref: String? = nil,
    title: String = "",
    readingOrderDuration: TimeInterval? = nil
  ) -> [FlutterMediaOverlay] {
    func flatten(_ obj: GuidedNavigationObject) -> [FlutterMediaOverlayItem] {
      var items: [FlutterMediaOverlayItem] = []
      // Accept textref (EPUB/read-aloud) or imgref (Divina panel audio) as the text anchor.
      if let audio = obj.audioref, let text = obj.textref ?? obj.imgref {
        items.append(FlutterMediaOverlayItem(
          audio: audio,
          text: text,
          position: position,
          tocTitle: title.isEmpty ? nil : title,
          tocHref: tocHref,
          readingOrderDuration: readingOrderDuration
        ))
      }
      for child in obj.children {
        items += flatten(child)
      }
      return items
    }

    let allItems = guided.flatMap { flatten($0) }

    // Group by (audioFile, textFile), preserving insertion order.
    var orderedKeys: [(String, String)] = []
    var itemsByKey: [String: [FlutterMediaOverlayItem]] = [:]
    for item in allItems {
      let key = "\(item.audioFile)|\(item.textFile)"
      if itemsByKey[key] == nil {
        orderedKeys.append((item.audioFile, item.textFile))
        itemsByKey[key] = []
      }
      itemsByKey[key]!.append(item)
    }

    return orderedKeys.compactMap { (audioFile, textFile) in
      let key = "\(audioFile)|\(textFile)"
      guard let items = itemsByKey[key], !items.isEmpty else { return nil }
      return FlutterMediaOverlay(items: items, readingOrderDuration: readingOrderDuration)
    }
  }
}
